### PR TITLE
[Optimization] Return image_grid_thw in render response for disaggregated mRoPE

### PR DESCRIPTION
**Pull request by roytman**
## Purpose

In disaggregated serving, `kwargs_data` contains serialized `pixel_values` tensors that dominate payload size. The prefill and decode workers only need `image_grid_thw` (a 3-integer-per-image array) for mRoPE -- the encoder already consumed the pixels on a separate node. This change makes it possible to omit the heavy blobs entirely.

## Summary

- Adds `image_grid_thw` as a separate lightweight JSON field in the render response (`MultiModalFeatures`), alongside the existing `kwargs_data` blobs.
- The generate endpoint now accepts `image_grid_thw` directly, constructing minimal `MultiModalKwargsItem` objects for mRoPE position computation without deserializing the full msgpack `kwargs_data`.
- In E+P+D (or any disaggregated setup where the encoder runs separately), the proxy can strip `kwargs_data` from the payload forwarded to prefill/decode nodes, significantly reducing transfer size for large images.

## Test Plan

- [x] Existing test `tests/entrypoints/serve/disagg/test_serving_multimodal_tokens.py` passes (validates `image_grid_thw` structure in render response)
- [ ] Manual E2E: render a multimodal request, strip `kwargs_data` from the response, send only `image_grid_thw` + `mm_hashes` + `mm_placeholders` to the generate endpoint on a node without encoder -- verify mRoPE positions computed correctly

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing a test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/entrypoints/serve/disagg/serving.py [hunk 1]
```diff
@@ -128,6 +131,23 @@ async def serve_tokens(
                         decode_mm_kwargs_item(item) if item is not None else None
                         for item in items
                     ]
+            elif features.image_grid_thw is not None:
+                # Lightweight path: construct minimal items containing
+                # only grid metadata for mRoPE position computation.
+                for modality, grids in features.image_grid_thw.items():
+                    items_list: list[MultiModalKwargsItem | None] = []
```
[the remark this answers is not in the document]
Comment (roytman) [reply vllm/entrypoints/serve/disagg/serving.py:166]: false positive
vLLM requires Python 3.10+ ([pyproject.toml](https://github.com/vllm-project/vllm/blob/c199cfe3286bb3504335aa15e782e330e6b60473/pyproject.toml): requires-python = ">=3.10,<3.15")

File: vllm/entrypoints/serve/disagg/protocol.py [hunk 2]
```diff
@@ -52,6 +52,15 @@ class MultiModalFeatures(BaseModel):
     ``None`` for metadata-only (cache-hit) responses.
     """
 
+    image_grid_thw: dict[str, list[list[int] | None]] | None = None
```
[the remark this answers is not in the document]
Comment (roytman) [reply vllm/entrypoints/serve/disagg/protocol.py:61]: false positive
vLLM requires Python 3.10+ ([pyproject.toml](https://github.com/vllm-project/vllm/blob/c199cfe3286bb3504335aa15e782e330e6b60473/pyproject.toml): requires-python = ">=3.10,<3.15")

File: vllm/entrypoints/serve/render/serving.py [hunk 3]
```diff
@@ -380,18 +383,33 @@ def _extract_mm_features(
...
                 kwargs_data[modality] = [
                     encode_mm_kwargs_item(item) if item is not None else None
                     for item in items
                 ]
+                thw_key = f"{modality}_grid_thw"
+                grids: list[list[int] | None] = []
+                for item in items:
+                    if item is not None and thw_key in item:
+                        thw_tensor = cast("torch.Tensor", item[thw_key].data)
+                        grids.append(thw_tensor.tolist())
```
[the remark this answers is not in the document]
Comment (roytman) [reply vllm/entrypoints/serve/render/serving.py:400]: Per-item image_grid_thw.data is already 1-D shape (3,), not (1, 3). Evidence: [MultiModalBatchedField.build_elems](https://github.com/vllm-project/vllm/blob/c199cfe3286bb3504335aa15e782e330e6b60473/vllm/multimodal/inputs.py#L470-L478) splits the batched tensor along dim 0 (so (N, 3) → N elems of shape (3,)), and Qwen2-VL itself unpacks per-item with t, h, w = mm_feature.data["image_grid_thw"].data.tolist() ([qwen2_vl.py:1217](https://github.com/vllm-project/vllm/blob/c199cfe3286bb3504335aa15e782e330e6b60473/vllm/model_executor/models/qwen2_vl.py#L1215-L1220)). So tolist() already returns a flat [t, h, w], which is exactly what your test asserts (len(grid) == 3, all ints). No .view(-1) needed

File: vllm/entrypoints/serve/disagg/serving.py [hunk 4]
```diff
@@ -128,6 +131,23 @@ async def serve_tokens(
...
+                    thw_key = f"{modality}_grid_thw"
+                    for grid in grids:
+                        if grid is not None:
+                            tensor = torch.tensor([grid], dtype=torch.int64)
```
[the remark this answers is not in the document]
Comment (roytman) [reply vllm/entrypoints/serve/disagg/serving.py]: Fixed in c199cfe3: now [torch.tensor(grid, dtype=torch.int64)](https://github.com/vllm-project/vllm/blob/c199cfe3286bb3504335aa15e782e330e6b60473/vllm/entrypoints/serve/disagg/serving.py#L168-L172)
